latextools_utils/analysis.py

```diff
@@ -0,0 +1,426 @@
+        if file_name not in self._content:
+            raise FileNotAnalyzed(file_name)
+        return self._content[file_name]
+
+    def raw_content(self, file_name):
+        """
+        The raw unprocessed content of the file (a string)
+        """
+        if file_name not in self._raw_content:
+            raise FileNotAnalyzed(file_name)
+        return self._raw_content[file_name]
+
+    def rowcol(self, file_name):
+        """
+        Returns a rowcol function for the file with the same behavior as the
+        view.rowcol function from the sublime api
+        """
+        return make_rowcol(self.raw_content(file_name))
+
+    def commands(self, flags=DEFAULT_FLAGS):
+        """
+        Returns a list with copies of each command entry in the document
+
+        Arguments:
+        flags -- flags to filter the commands, which should for a be used over
+            over filtering the commands on your own (optimization/caching).
+            Possible flags are:
+            NO_BEGIN_END_COMMANDS - removes all begind and end commands
+                i.e.: exclude \\begin{} and \\end{} commands
+            ONLY_COMMANDS_WITH_ARGS - removes all commands without arguments
+                i.e.: allows \\com{args} and \\com{} but not \\com
+            default flags are:
+            NO_BEGIN_END_COMMANDS | ONLY_COMMANDS_WITH_ARGS
+
+        Returns:
+        A list of all commands, which are preprocessed with the flags
+        """
+        return _copy_entries(self._commands(flags))
+
+    def filter_commands(self, how, flags=DEFAULT_FLAGS):
+        """
+        Returns a filtered list with copies of each command entry
+        in the document
+
+        Arguments:
+        how -- how it should be filtered, possible types are:
+            string - only commands, which equals this string
+            list of string - only commands which are in this list
+            function of string->bool - should return true iff the command
+                should be in the result
+        flags -- flags to filter the commands, which should for a be used over
+            over filtering the commands on your own (optimization/caching).
+            Possible flags are:
+            NO_BEGIN_END_COMMANDS - removes all begin and end commands
+                i.e.: exclude \\begin{} and \\end{} commands
+            ONLY_COMMANDS_WITH_ARGS - removes all commands without arguments
+                i.e.: allows \\com{args} and \\com{} but not \\com
+            default flags are:
+            NO_BEGIN_END_COMMANDS | ONLY_COMMANDS_WITH_ARGS
+
+        Returns:
+        A list of all commands, which are preprocessed with the flags
+        """
+        # convert the filter into a function
+        if type(how) is str:
+            def command_filter(c): return c.command == how
+        elif type(how) is list:
+            def command_filter(c): return c.command in how
+        elif callable(how):
+            def command_filter(c): return how(c.command)
```

Comment:
Yes I agree and will change this.

My inital decision not to pass the whole command was based on the fear, that the caller could change the command content before creating a copy. Nevertheless I don't think that is a big issue, since the caching process is done before the analysis is returned and hence everyone else gets a "fresh" analysis. In addition the attributes are all protected and not private anyway.